## 1.2 Yoğunluk Matrisleri (Density matrices)

Kuantum durumlarını ifade etmek için çok daha genel ve kapsayıcı bir matematiksel yöntem olan **yoğunluk matrislerini** inceleyerek işe başlayalım. 
İlk olarak, birden fazla kübit barındıran bir sistemde, bu kübitlerden sadece bir tanesinin durumunu kendi başına nasıl yazabileceğimiz sorusuyla başlayalım. Bu amaçla, $A$ ve $B$ adında iki kuantum sistemimiz olduğunu hayal edelim. Örneğin $A$ ve $B$, birleşik bir $|\Psi\rangle_{AB}$ durumunda bulunan iki kübit olsun ve biz sadece $A$ kübitinin durumunu bilmek isteyelim. 

Eğer bu birleşik durum bir çarpım durumu (product state) ise; yani $|\Psi\rangle_{AB} = |\psi\rangle_A \otimes |\psi\rangle_B$ şeklinde, $A$'nın $|\psi\rangle_A$ durumu ile $B$'nin $|\psi\rangle_B$ durumunun tensör çarpımı alınarak elde edilmişse cevap oldukça açıktır: $A$ kübitinin durumu basitçe $|\psi\rangle_A$ vektörüdür.


Ancak, Bölüm00 notlarından hatırlayabileceğiniz üzere, $A$ and $B$ sistemleri üzerinde tanımlı bazı iki-parçalı (bipartite) kuantum durumları, tensör çarpımlarının bir süperpozisyonu olarak yazılabilir. Bu durum, bileşke sistemin tek bir tensör çarpımı şeklinde yazılıp yazılamayacağını belirsiz hale getirir. 

Bunun en güzel örneği, daha önce gördüğümüz meşhur **EPR çiftidir**:

$$|\text{EPR}\rangle_{AB} = \frac{1}{\sqrt{2}}|0\rangle_A|0\rangle_B + \frac{1}{\sqrt{2}}|1\rangle_A|1\rangle_B$$

Bu tür dolanık (entangled) durumlar için birleşik sistemi $|\Psi\rangle_{AB} = |\psi\rangle_A \otimes |\psi\rangle_B$ şeklinde iki bağımsız vektörün çarpımı olarak **ifade edemeyiz**. Dolayısıyla, $B$ sistemine hiçbir atıfta bulunmadan (referans vermeden) sadece $A$ kübitinin durumunu nasıl tek başına yazabileceğimiz belirsizleşir. 

Yine de böyle bir tanım matematiksel olarak mümkün olmalıdır; ne de olsa o parçacık fiziksel olarak varlığını sürdürmektedir! Eğer mevcut ket/bra vektör gösterimimiz bu durumu tanımlamaya yetmiyorsa, bu durum kuantum formalizmimizin henüz eksik olduğu anlamına gelir. Bu yüzden daha genel matematiksel bir genişletmeye (genelleştirmeye) ihtiyaç duyarız.


Daha genel bir kuantum durumu tanımına ihtiyaç duymamızın ikinci motivasyonu ise, olasılıksal bir sürecin (örneğin bir ölçüm işleminin) belirli olasılıklarla farklı kuantum durumları ürettiği durumlardan kaynaklanır. 

Örneğin, elimizde $p_1$ olasılığıyla üretilen bir $|\psi_1\rangle$ durumu veya $p_2$ olasılığıyla üretilen başka bir $|\psi_2\rangle$ durumu olduğunu varsayalım. Bu kuantum durumunu eksiksiz ve doğru bir şekilde ifade edebilmek için hem olası durumları hem de onların gerçekleşme olasılıklarını ($\{|\psi_i\rangle, p_i\}$) aynı anda hesaba katmak zorundayız. 

Peki, böyle bir olasılıksal süreç tarafından yaratılan kuantum durumunun uygun matematiksel tanımını tek bir çatı altında nasıl yazabiliriz?


### 1.2.1 Giriş 

Önceki bölümde sorduğumuz soruların cevabı, **yoğunluk matrisi formalizmi** olarak adlandırılan yapıda yatmaktadır. Başlangıç olarak, tek bir sistemin kuantum durumunu ($|\psi\rangle$), ket ve bra vektörlerinin dış çarpımı (outer product) şeklinde bir matris olarak yazalım:

$$\rho = |\psi\rangle\langle\psi|$$

Bu matrisin **rank-1** (mertebesi 1 olan) bir matris olduğuna dikkat edin. Bu matris, tam olarak 1 adet sıfırdan farklı özdeğere (eigenvalue) sahiptir (bu özdeğer 1'e eşittir) ve bu özdeğere karşılık gelen özdurum (eigenstate) ise $|\psi\rangle$ vektörünün kendisidir. Bu şekilde tek bir vektörle yazılabilen durumlara **saf durumlar (pure states)** denir.


> 📝 **Örnek 1.2.1 ** $|1\rangle$ ve $|-\rangle = (|0\rangle - |1\rangle)/\sqrt{2}$ kuantum durumlarına karşılık gelen yoğunluk matrislerini dış çarpım yöntemiyle inşa edelim:
> 
> $$|1\rangle\langle1| = \begin{pmatrix} 0 \\ 1 \end{pmatrix} \begin{pmatrix} 0 & 1 \end{pmatrix} = \begin{pmatrix} 0 & 0 \\ 0 & 1 \end{pmatrix} \quad \text{--- (1.3)}$$
> 
> $$|-\rangle\langle-| = \frac{1}{2} \begin{pmatrix} 1 \\ -1 \end{pmatrix} \begin{pmatrix} 1 & -1 \end{pmatrix} = \frac{1}{2} \begin{pmatrix} 1 & -1 \\ -1 & 1 \end{pmatrix} \quad \text{--- (1.4)}$$


In [1]:
import numpy as np

# Temel kuantum ket vektörleri
qubit_0 = np.array([[1], [0]]) # |0>
qubit_1 = np.array([[0], [1]]) # |1>

# |-> durumunu oluşturalım: (|0> - |1>) / sqrt(2)
minus_ket = (qubit_0 - qubit_1) / np.sqrt(2)

# Dış çarpım (outer product) yöntemiyle yoğunluk matrislerini (rho) hesaplayalım
rho_1 = np.dot(qubit_1, qubit_1.T)
rho_minus = np.dot(minus_ket, minus_ket.T)

print("|1><1| Yoğunluk Matrisi:\n", rho_1)
print("\n|-><-| Yoğunluk Matrisi:\n", rho_minus)


|1><1| Yoğunluk Matrisi:
 [[0 0]
 [0 1]]

|-><-| Yoğunluk Matrisi:
 [[ 0.5 -0.5]
 [-0.5  0.5]]


Durumları matris olarak yazmak, yukarıda bahsettiğimiz problemleri çözmemize nasıl yardımcı olur?

Bir laboratuvarda bir uzmanın, $p_1 = p_2 = 1/2$ eşit olasılıklarıyla iki olası durumdan ($|\psi_1\rangle$ veya $|\psi_2\rangle$) birini hazırladığı senaryoyu düşünelim. Bu senaryoda durumların basit bir süperpozisyonunu (üst üste binmesini) yazmak kesinlikle **yanlış bir tanım** olacaktır. Çünkü parçacık aslında bu iki durumdan tam olarak bir tanesindedir; sadece biz hangisi olduğunu bilmiyoruzdur (klasik bir belirsizlik söz konusudur). 

Eğer hazırlanan sistemin kimliği bizim tarafımızdan bilinmiyorsa, ortaya çıkan nihai durumu nasıl ifade edebiliriz? İşte bu durumda, sistemi $|\psi_1\rangle$ ve $|\psi_2\rangle$ durumlarının istatistiksel bir **karışımı (mixture)** olarak tanımlayabiliriz.

Eşit olasılıklı bu senaryo için elde edilen birleşik karma durum şu şekildedir:

$$\rho = \frac{1}{2}|\psi_1\rangle\langle\psi_1| + \frac{1}{2}|\psi_2\rangle\langle\psi_2| \quad \text{--- (1.5)}$$

Bu şekilde elde edilen $\rho$ matrisine genel olarak bir **yoğunluk matrisi (density matrix)** denir. Genel tanımı genişletirsek; eğer bir kaynak, bir $|\psi_x\rangle$ kuantum durumunu $p_x$ olasılığı ile hazırlıyorsa, oluşan nihai sistemin yoğunluk matrisi şu formülü alır:

$$\rho = \sum_{x} p_x |\psi_x\rangle\langle\psi_x| \quad \text{--- (1.6)}$$


Peki bu tanım neden bu kadar başarılı ve doğrudur? Sistemyi standart hesapsal tabanda ölçtüğümüzde ne olacağına bakarak bunu doğrulayalım. Eğer sistemimiz kesin olarak tek bir $|\psi_j\rangle$ durumunda olsaydı, ölçüm sonuçlarının olasılıklarının (Born kuralı gereği) şu şekilde olmasını beklerdik:

$$q_{0|j} = |\langle 0 | \psi_j \rangle|^2 = \langle 0 | \psi_j \rangle \langle \psi_j | 0 \rangle \quad \text{--- (1.7)}$$

$$q_{1|j} = |\langle 1 | \psi_j \rangle|^2 = \langle 1 | \psi_j \rangle \langle \psi_j | 1 \rangle \quad \text{--- (1.8)}$$

Ancak, her bir $|\psi_j\rangle$ durumu sisteme $p_j$ olasılığı ile dahil ediliyorsa, toplam olasılık teorisi gereği nihai ölçümde '0' ve '1' görme olasılıklarımızın şu şekilde ağırlıklı toplanması gerekir:

$$q_0 = \sum_{j} p_j q_{0|j} \quad \text{--- (1.9)}$$

$$q_1 = \sum_{j} p_j q_{1|j} \quad \text{--- (1.10)}$$


Şimdi bu terimlerden ilkini ($q_0$) açarak yoğunluk matrisi formalizmiyle olan muazzam ilişkisini ortaya çıkaralım:

$$q_0 = \sum_{j} p_j q_{0|j} = \sum_{j} p_j \langle 0 | \psi_j \rangle \langle \psi_j | 0 \rangle = \langle 0 | \left( \sum_{j} p_j |\psi_j\rangle\langle \psi_j| \right) | 0 \rangle = \langle 0 | \rho | 0 \rangle \quad \text{--- (1.11)}$$

Görüldüğü üzere, karmaşık olasılık adımlarının sonucu doğrudan $\langle 0 | \rho | 0 \rangle$ sandviç çarpımına eşit çıkmaktadır. Dolayısıyla, $\rho$ yoğunluk matrisi, ölçüm sonuçlarından sezgisel olarak beklediğimiz tüm istatistiksel olasılıkları kendi içinde kusursuz bir şekilde barındırır ve yansıtır.


> 📝 **Örnek 1.2.2** Eğer bir kaynak kuantum durumlarını olasılıksal bir yaklaşımla hazırlıyorsa; yani $p_x$ olasılığıyla bir $\rho_x$ yoğunluk matrisi üretiyorsa, ortaya çıkan nihai bileşke yoğunluk matrisi şu ağırlıklı toplam formülü ile hesaplanır:
> 
> $$\rho = \sum_{x} p_x \rho_x \quad \text{--- (1.12)}$$
> 
> Olasılıklar ve yoğunluk matrislerinden oluşan bu $\mathcal{E} = \{(p_x, \rho_x)\}_x$ kümesine kuantum literatüründe bir **durum ansamblı (ensemble of states)** adı verilir. Kaynağın sadece saf (pure) durumlar hazırladığı senaryonun, tek bir $x$ değeri için $p_x = 1$ ve $\rho_x = |\psi_x\rangle\langle\psi_x|$ olan özel bir durum ansamblına karşılık geldiğine dikkat edilmelidir.


> 📝 **Örnek 1.2.3 ** Bir kuantum kaynağının $1/3$ olasılıkla $|1\rangle\langle1|$ durumunu ve $2/3$ olasılıkla $|+\rangle\langle+|$ durumunu hazırladığını varsayalım. Bu durumda $\mathcal{E} = \{(1/3, |1\rangle\langle1|), (2/3, |+\rangle\langle+|)\}$ ansamblı için nihai yoğunluk matrisi şu şekilde hesaplanır:
> 
> $$\rho = \frac{1}{3}|1\rangle\langle1| + \frac{2}{3}|+\rangle\langle+|$$
> 
> Önceki bölümlerde elde ettiğimiz matris gösterimlerini yerine koyarsak:
> 
> $$\rho = \frac{1}{3} \begin{pmatrix} 0 & 0 \\ 0 & 1 \end{pmatrix} + \frac{2}{3} \cdot \frac{1}{2} \begin{pmatrix} 1 & 1 \\ 1 & 1 \end{pmatrix}$$
> 
> $$\rho = \begin{pmatrix} 0 & 0 \\ 0 & 1/3 \end{pmatrix} + \begin{pmatrix} 1/3 & 1/3 \\ 1/3 & 1/3 \end{pmatrix} = \begin{pmatrix} 1/3 & 1/3 \\ 1/3 & 4/3 \end{pmatrix} = \frac{1}{3} \begin{pmatrix} 1 & 1 \\ 1 & 4 \end{pmatrix} \quad \text{--- (1.13)}$$


In [2]:
import numpy as np

# Bileşen matrisleri tanımlayalım
rho_1 = np.array([[0, 0], 
                  [0, 1]])

rho_plus = np.array([[0.5, 0.5], 
                     [0.5, 0.5]])

# Özgün olasılık ağırlıkları (1/3 ve 2/3)
p1 = 1/3
p2 = 2/3

# Ansambl toplam yoğunluk matrisi (rho)
rho_final = (p1 * rho_1) + (p2 * rho_plus)

print("Bilgisayar tarafından hesaplanan nihai karma matris:\n", rho_final)
print("\nTeorik matrisimiz ile uyuşuyor mu? (1/3 parantezinde [1,1],[1,4]):\n", np.array([[1, 1], [1, 4]]) / 3)


Bilgisayar tarafından hesaplanan nihai karma matris:
 [[0.33333333 0.33333333]
 [0.33333333 0.66666667]]

Teorik matrisimiz ile uyuşuyor mu? (1/3 parantezinde [1,1],[1,4]):
 [[0.33333333 0.33333333]
 [0.33333333 1.33333333]]


> 📝 **Örnek 1.2.4** Kuantumda **süperpozisyon (superposition)** ile **karışım (mixture)** kavramları kesinlikle aynı şey değildir. Sezgisel olarak aralarındaki fark şudur: Bir karışım, doğası gereği klasik bir belirsizliktir; yani arka planda sistemi belirli olasılıklarla ya bir durumda ya da diğer durumda hazırlayan fiziksel bir süreç vardır. Buna karşılık, süperpozisyondaki bir kuantum durumu ise aynı anda hem bir durumdur hem de diğeridir.
> 
> Bu derin farkı daha net görmek için $|0\rangle$ ve $|1\rangle$ durumlarını karıştırmak ile bir süperpozisyon oluşturmak arasındaki senaryoyu inceleyelim. Öncelikle, $p_0 = p_1 = 1/2$ olasılıklarıyla $|0\rangle$ ve $|1\rangle$ durumlarını hazırlayan olasılıksal bir kaynağı (karışımı) ele alalım. Bu karışımın yoğunluk matrisi şu şekilde bir birim matris ($\mathbb{I}/2$) üretir:
> 
> $$\rho = \frac{1}{2}|0\rangle\langle0| + \frac{1}{2}|1\rangle\langle1| = \frac{\mathbb{I}}{2} \quad \text{--- (1.14)}$$
> 
> Burada birim matrisimiz $\mathbb{I}$ şöyledir:
> 
> $$\mathbb{I} = \begin{pmatrix} 1 & 0 \\ 0 & 1 \end{pmatrix} \quad \text{--- (1.15)}$$


Şimdi, bu karma durumu (karışımı) alıp hesapsal taban yerine $\mathcal{H} = \{|+\rangle, |-\rangle\}$ **Hadamard tabanında** ölçtüğümüzü varsayalım. Taban elemanlarımızı hatırlayalım:
*   $|+\rangle = \frac{1}{\sqrt{2}}(|0\rangle + |1\rangle) \quad \text{--- (1.16)}$
*   $|-\rangle = \frac{1}{\sqrt{2}}(|0\rangle - |1\rangle) \quad \text{--- (1.17)}$

Bu ölçüm sonucunda $+$ ve $-$ görme olasılıklarımızı yoğunluk matrisini kullanarak hesaplarsak şu değerleri elde ederiz:

$$q_+ = \langle + | \rho | + \rangle = \langle + | \frac{\mathbb{I}}{2} | + \rangle = \frac{1}{2}\langle + | + \rangle = \frac{1}{2} \quad \text{--- (1.18)}$$

$$q_- = \langle - | \rho | - \rangle = \langle - | \frac{\mathbb{I}}{2} | - \rangle = \frac{1}{2}\langle - | - \rangle = \frac{1}{2} \quad \text{--- (1.19)}$$

Görüldüğü üzere, karışım durumunu Hadamard tabanında ölçtüğümüzde de sonuç yine yazı-tura gibi yarı yarıya rastgele ($1/2$) çıkmaktadır.


Buna tamamen zıt bir senaryo olarak, şimdi de doğrudan kuantum süperpozisyonu olan $|+\rangle$ saf durumunu ele alalım. 

Eğer saf $|+\rangle$ durumunu Hadamard tabanında ölçersek, sonuçlar **$q_+ = 1$** ve **$q_- = 0$** olarak elde edilir. Bu durum, kuantum süperpozisyonları ile klasik karışımlar arasındaki en temel ve köklü farkı açıkça ortaya koymaktadır. Klasik karışımlardaki belirsizlikten taban değiştirerek kurtulamazken, kuantum süperpozisyonlarında doğru tabanı seçtiğimizde belirsizlik tamamen ortadan kalkar ve kesinlik elde edilir.


### 1.2.2 Bazı Matematiksel Tanımlar (Some mathematical definitions)

Yoğunluk matrislerini ve özelliklerini resmi olarak tanımlamak için lineer cebirden bazı önemli kavramları hatırlayalım. Tanıtacağımız ilk kavram, bizim bağlamımızda matrisleri ifade etmek için **lineer operatör (doğrusal operatör)** kavramıdır. Bu terim, bir matrisin bir vektörü alıp başka bir vektöre eşlediği (dönüştürdüğü) fikrini vurgular. Matris çarpımı doğrusal olduğundan, bu işlem de doğrusal bir işlem olacaktır. Bu nedenle kuantum literatüründe "matris" ve "operatör" terimlerini birbirinin yerine kullanacağız.

> 🟦 **Tanım 1.2.1 — Lineer Operatör.** $d$-boyutlu karmaşık bir $\mathbb{C}^d$ vektör uzayını ele alalım. Bir $L: \mathbb{C}^d \rightarrow \mathbb{C}^{d'}$ lineer operatörü, $d' \times d$ boyutunda bir matris olarak temsil edilebilir:
> 
> $$L = \begin{pmatrix} L_{11} & L_{12} & \dots & L_{1d} \\ L_{21} & \ddots & \dots & L_{2d} \\ \vdots & \vdots & \ddots & \vdots \\ L_{d'1} & L_{d'2} & \dots & L_{d'd} \end{pmatrix} \quad \text{--- (1.22)}$$
> 
> Burada her bir eleman $L_{ij} \in \mathbb{C}$ karmaşık sayısıdır. Tüm lineer operatörlerin kümesi $\mathcal{L}(\mathbb{C}^d, \mathbb{C}^{d'})$ ile gösterilir.


> 🟦 **Tanım 1.2.2 — Hermityen Matris.** Bir $M \in \mathcal{L}(\mathbb{C}^d, \mathbb{C}^d)$ lineer operatörü, eğer kendi eşlenik transpozuna (adjoint) eşitse, yani **$M^\dagger = M$** ise **Hermityen** olarak adlandırılır.

**Spektral Teorem (Spectral Theorem)**, herhangi bir Hermityen $M$ operatörünün, reel özdeğerler (real eigenvalues) ile köşegenleştirilebileceğini söyler. Bu durum, $\mathbb{C}^d$ uzayının bir ortonormal tabanını oluşturan özdurumlar (özvektörler) $\{|v_j\rangle\}$ ve reel $\lambda_j$ özdeğerleri için $M$ matrisinin şu şekilde yazılabileceği anlamına gelir:
$$M = \sum_{j} \lambda_j |v_j\rangle\langle v_j|$$

> 🟦 **Tanım 1.2.3 — Pozitif Yarı-Tanımlı Matris (Positive Semidefinite).** Hermityen bir $M$ matrisinin tüm özdeğerleri ($\{\lambda_i\}_i$) negatif değilse, yani **$\lambda_i \geq 0$** ise bu matris **pozitif yarı-tanımlıdır**. Bu durum kuantum literatüründe genellikle **$M \geq 0$** şeklinde gösterilir.


> 🟦 **Alıştırma 1.2.2 (Çözümlü Versiyon)** Bir $M$ matrisinin pozitif yarı-tanımlı olması için gerek ve yeter şartın, tüm birim $|v\rangle$ vektörleri için $\langle v|M|v\rangle \geq 0$ olduğunu gösteriniz. Özel olarak, $M$ matrisinin herhangi bir tabandaki köşegen elemanlarının ($\langle i|M|i\rangle$) negatif olamayacağını gösteriniz. Bunun **yeterli bir şart olmadığını** ispatlamak adına: Köşegen elemanları pozitif olan fakat kendisi pozitif yarı-tanımlı olmayan bir $M$ matrisi örneği bulunuz.

**Çözüm:**
Köşegen elemanlarının pozitif olması, bir matrisin pozitif yarı-tanımlı olması için tek başına yeterli değildir. Bunu görmek için aşağıdaki Hermityen matrisi ele alalım:

$$M = \begin{pmatrix} 1 & 2 \\ 2 & 1 \end{pmatrix}$$

Bu matrisin köşegen elemanları $M_{11} = 1$ ve $M_{22} = 1$ olup, ikisi de pozitiftir ($>0$). Şimdi bu matrisin özdeğerlerini ($\lambda$) bulalım:
$$\det(M - \lambda\mathbb{I}) = \det\begin{pmatrix} 1-\lambda & 2 \\ 2 & 1-\lambda \end{pmatrix} = (1-\lambda)^2 - 4 = 0$$
$$\lambda^2 - 2\lambda - 3 = 0 \implies (\lambda - 3)(\lambda + 1) = 0$$

Buradan özdeğerler **$\lambda_1 = 3$** ve **$\lambda_2 = -1$** olarak bulunur. Özdeğerlerden biri negatif ($\lambda_2 = -1 < 0$) çıktığı için, köşegenleri pozitif olmasına rağmen bu matris **pozitif yarı-tanımlı değildir**. Bu durum alıştırmadaki iddiayı somut olarak ispatlar.


Matrisler üzerindeki bir diğer önemli işlem ise **iz (trace)** işlemidir. Bölüm00 notlarında gördüğümüz gibi, bir matrisin izi basitçe onun köşegen elemanlarının toplamıdır. Bu yönüyle iz işlemi, herhangi bir matrisi alıp bir karmaşık sayıya eşleyen doğrusal bir dönüşümdür. 


In [1]:
import numpy as np

# Karşıt örnek matrisimizi tanımlayalım
M = np.array([[1, 2], 
              [2, 1]])

# Matrisin özdeğerlerini hesaplayalım
ozdegerler = np.linalg.eigvals(M)

print("Matrisin Köşegen Elemanları:", np.diag(M))
print("Matrisin Özdeğerleri:", ozdegerler)

# Pozitif yarı-tanımlılık kontrolü (Tüm özdeğerler >= 0 mı?)
is_psd = np.all(ozdegerler >= 0)
print(f"Bu matris pozitif yarı-tanımlı mı?: {is_psd}")


Matrisin Köşegen Elemanları: [1 1]
Matrisin Özdeğerleri: [ 3. -1.]
Bu matris pozitif yarı-tanımlı mı?: False


> 🟦 **Tanım 1.2.4 — Matris İzi (Trace of a matrix).** Bir $M \in \mathcal{L}(\mathbb{C}^d, \mathbb{C}^d)$ matrisinin izi, aşağıdaki formül ile tanımlanır:
> 
> $$\text{tr}(M) = \sum_{i} \langle i|M|i\rangle$$
> 
> Burada $\{|i\rangle\}$, $\mathbb{C}^d$ uzayının herhangi bir ortonormal tabanıdır.

Matris izinin değerinin, seçilen ortonormal tabandan tamamen bağımsız olduğunu ve hangi tabanı seçerseniz seçin sonucun değişmeyeceğini kendiniz de görebilirsiniz. İzin en önemli özelliklerinden biri ise **döngüsel (cyclic)** olmasıdır.


> 🟦 **Alıştırma 1.2.3** Herhangi $M, N$ matrisleri için $\text{tr}(MN) = \text{tr}(NM)$ olduğunu gösteriniz. 

Kuantum hesaplama literatüründe bu döngüsel özelliği, formüller üzerinde şu tür matematiksel manipülasyonlar gerçekleştirmek için çok sık kullanırız:
$$\langle i|M|i\rangle = \text{tr}(\langle i|M|i\rangle) = \text{tr}(M|i\rangle\langle i|) \quad \text{--- (1.23)}$$

Ancak, matrislerin döngüsel olmayan (non-cyclic) bir permütasyonunun (sıralamasının) genel olarak iz değerini **korumadığına** dikkat etmek gerekir. Daha açık bir ifadeyle, $M, N, P$ matrisleri için genel olarak durum şöyledir:
$$\text{tr}(MNP) \neq \text{tr}(NMP) \quad \text{--- (1.24)}$$


> 📝 **Alıştırma 1.2.3 İçin Somut Karşıt Örnek:** Denklem (1.24)'te iddia edilen, döngüsel olmayan sıralamanın izi bozduğunu ispatlamak için Pauli matrislerimizi ($X, Y, Z$) kullanalım:
> 
> *   $M = X = \begin{pmatrix} 0 & 1 \\ 1 & 0 \end{pmatrix}$
> *   $N = Y = \begin{pmatrix} 0 & -i \\ i & 0 \end{pmatrix}$
> *   $P = Z = \begin{pmatrix} 1 & 0 \\ 0 & -1 \end{pmatrix}$
> 
> 1. **Döngüsel Sıralama ($XYZ$):** Pauli matrislerinin özelliğinden $XYZ = i\mathbb{I}$ olduğunu biliyoruz.
>    $$\text{tr}(XYZ) = \text{tr}\begin{pmatrix} i & 0 \\ 0 & i \end{pmatrix} = i + i = 2i$$
> 
> 2. **Döngüsel Olmayan Sıralama ($XZY$):** Çarpımı hesaplayalım: $XZ = -iY$ ve $(-iY)Y = -i\mathbb{I}$ olur.
>    $$\text{tr}(XZY) = \text{tr}\begin{pmatrix} -i & 0 \\ 0 & -i \end{pmatrix} = -i - i = -2i$$
> 
> Görüldüğü üzere, $\text{tr}(XYZ) \neq \text{tr}(XZY)$ elde edilir ($2i \neq -2i$). Bu durum, matrislerin yerini rastgele değiştirmenin izi bozduğunu somut olarak kanıtlar.


In [4]:
import numpy as np

# Pauli matrislerini doğru şekilde tanımlayalım
X = np.array([[0, 1], [1, 0]])
Y = np.array([[0, -1j], [1j, 0]])
Z = np.array([[1, 0], [0, -1]])

# 1. tr(MNP) hesabını yapalım (X * Y * Z)
XYZ = np.linalg.multi_dot([X, Y, Z])
tr_XYZ = np.trace(XYZ)

# 2. tr(NMP) hesabını yapalım (X * Z * Y) -> Döngüsel olmayan sıra
XZY = np.linalg.multi_dot([X, Z, Y])
tr_XZY = np.trace(XZY)

print("tr(XYZ) Sonucu (Döngüsel):", tr_XYZ)
print("tr(XZY) Sonucu (Döngüsel Olmayan):", tr_XZY)
print("İki iz değeri birbirine eşit mi?:", np.isclose(tr_XYZ, tr_XZY))




tr(XYZ) Sonucu (Döngüsel): 2j
tr(XZY) Sonucu (Döngüsel Olmayan): -2j
İki iz değeri birbirine eşit mi?: False


### 1.2.3 Yoğunluk Matrisleri ve Özellikleri (Density matrices and their properties)

Yukarıdaki tartışmalardan yola çıkarak, kuantum durumlarını daha genel bir çerçevede tanımlamak için $\rho$ yoğunluk matrisini kullanmaya karar vermiştik. Şimdi, rastgele bir $\rho$ matrisinin hangi durumlarda **geçerli bir yoğunluk matrisi** (yani gerçek bir kuantum durumu) kabul edileceğini biçimsel olarak inceleyelim.

Bir matrisin geçerli bir kuantum durumunu temsil edebilmesi için **gerek ve yeter olan iki temel özellik** mevcuttur: Matris pozitif yarı-tanımlı olmalı ve izi (trace) 1'e eşit olmalıdır.

Bunun neden doğru olduğunu görmek için, bir $\rho$ yoğunluk matrisinin $\{\lambda_j\}$ özdeğerleri ve bunlara karşılık gelen $\{|v_j\rangle\}$ özvektörleri cinsinden köşegenleştirilmiş ifadesini ele alalım:

$$\rho = \sum_{j} \lambda_j |v_j\rangle\langle v_j| \quad \text{--- (1.25)}$$

Burada $\{|v_j\rangle\}$ vektörleri ortonormaldir. Şimdi bu $\rho$ durumunu başka bir ortonormal $\{|w_k\rangle\}$ tabanında ölçtüğümüzü hayal edelim. Sistemin $\lambda_j$ olasılığıyla $|v_j\rangle\langle v_j|$ durumunda hazırlandığı bilgisinden yola çıkarak; sistem bu durumdayken yapılan ölçümde $k$ sonucunu elde etme koşullu olasılığı ($q_{k|j}$) Born kuralı ile şu şekilde verilir:

$$q_{k|j} = |\langle v_j | w_k \rangle|^2 = \langle w_k | v_j \rangle\langle v_j | w_k \rangle \quad \text{--- (1.26)}$$

Buna göre, $\rho$ durumu ölçüldüğünde genel olarak $k$ sonucunu elde etme toplam olasılığımız ($q_k$) şu ağırlıklı toplam formülüne dönüşür:

$$q_k = \sum_{j} \lambda_j q_{k|j} = \langle w_k | \left( \sum_{j} \lambda_j |v_j\rangle\langle v_j| \right) | w_k \rangle = \langle w_k | \rho | w_k \rangle \quad \text{--- (1.27)}$$

Olasılık teorisi gereği $q_k \geq 0$ ve $\sum_k q_k = 1$ olmak zorundadır. Eğer ölçümü matrisin kendi özdurum tabanında ($|w_j\rangle = |v_j\rangle$) yaptığımızı hayal edersek, doğrudan **$\lambda_j \geq 0$** sonucuna ulaşırız; yani $\rho$ matrisi **pozitif yarı-tanımlı** olmalıdır. Benzer şekilde, olasılıkların toplamı 1 olmak zorunda olduğundan:

$$1 = \sum_{j} q_j = \sum_{j} \lambda_j \text{tr}(|v_j\rangle\langle v_j|) = \text{tr}(\rho) \quad \text{--- (1.28)}$$

elde edilir. Bu da matrisin izinin (köşegen toplamının) neden tam olarak 1 olması gerektiğini ispatlar.


Bu çıkarımlar bizi bir kuantum sisteminin durumunu tanımlamanın en genel yolu olan şu resmi tanıma ulaştırır:

> 🟦 **Tanım 1.2.5 — Yoğunluk Matrisi (Density Matrix).** $\mathbb{C}^d$ durum uzayına sahip bir kuantum sistemini ele alalım. Genellikle $\rho$ ile gösterilen bir yoğunluk matrisi, aşağıdaki iki şartı sağlayan bir $\rho \in \mathcal{L}(\mathbb{C}^d, \mathbb{C}^d)$ lineer operatörüdür:
> 1. **$\rho \geq 0$** (Pozitif yarı-tanımlıdır, tüm özdeğerleri $\geq 0$'dır).
> 2. **$\text{tr}(\rho) = 1$** (Matrisin izi 1'dir).
> 
> Ek olarak; eğer matrisin mertebesi (rank) 1 ise, yani **$\text{rank}(\rho) = 1$** ise $\rho$ bir **saf durum (pure state)** olarak adlandırılır. Aksi takdirde ($\text{rank}(\rho) > 1$), sistem bir **karma durumdur (mixed state)**.


> 📝 **Örnek Soru:** Aşağıdaki matrisin geçerli bir kuantum yoğunluk matrisi olup olmadığını, eğer geçerliyse saf mı yoksa karma bir durum mu olduğunu belirleyiniz:
> 
> $$\rho_{test} = \begin{pmatrix} 0.6 & 0.2 \\ 0.2 & 0.4 \end{pmatrix}$$
> 
> **Çözüm:**
> 1. **İz Kontrolü:** $\text{tr}(\rho_{test}) = 0.6 + 0.4 = 1.0$. (İlk şart sağlandı).
> 2. **Pozitif Yarı-Tanımlılık (Özdeğerler):** Köşegenler pozitif ve matrisin determinantı $\det(\rho) = (0.6 \cdot 0.4) - (0.2 \cdot 0.2) = 0.24 - 0.04 = 0.20 > 0$'dır. Determinant ve iz pozitif olduğundan tüm özdeğerler pozitiftir ($\rho \geq 0$). (İkinci şart sağlandı).
> 
> Matris iki şartı da sağladığı için **geçerli bir yoğunluk matrisidir**. İki özdeğeri de sıfırdan farklı olduğu için ($\text{rank}=2$) bu bir **karma durumdur (mixed state)**.


Son olarak, yoğunluk matrisi ile verilen bir sistem üzerinde yapılacak ölçümlerin olasılık kuralını özetleyelim:

> 🟦 **Tanım 1.2.6 — Yoğunluk Matrisinin Bir Tabanda Ölçülmesi.** $\rho$ durumundaki bir kuantum sistemini ele alalım. Bu sistemin $\{|b_j\rangle\}_j$ ortonormal tabanında ölçülmesi durumunda, $j$ sonucunu elde etme olasılığı ($q_j$) şu formülle hesaplanır:
> 
> $$q_j = \langle b_j | \rho | b_j \rangle \quad \text{--- (1.29)}$$


In [5]:
import numpy as np

# Test yoğunluk matrisimizi tanımlayalım
rho = np.array([[0.6, 0.2], 
                [0.2, 0.4]])

# 1. İz kontrolü
iz = np.trace(rho)
print(f"Matrisin İzi: {iz} (Geçerli mi?: {np.isclose(iz, 1.0)})")

# 2. Özdeğer kontrolü (Pozitif yarı-tanımlılık)
ozdegerler = np.linalg.eigvals(rho)
print(f"Matrisin Özdeğerleri: {ozdegerler}")
is_psd = np.all(ozdegerler >= -1e-9) # Sayısal hassasiyet payı ile >= 0 kontrolü
print(f"Pozitif Yarı-Tanımlı mı?: {is_psd}")

# 3. Durumun türü (Saf mı Karma mı?)
rank = np.linalg.matrix_rank(rho)
durum_turu = "Saf Durum (Pure)" if rank == 1 else "Karma Durum (Mixed)"
print(f"Sistemin Türü: {durum_turu} (Rank: {rank})")


Matrisin İzi: 1.0 (Geçerli mi?: True)
Matrisin Özdeğerleri: [0.7236068 0.2763932]
Pozitif Yarı-Tanımlı mı?: True
Sistemin Türü: Karma Durum (Mixed) (Rank: 2)


### 1.2.4 Tek Kübit Karma Durumlar İçin Bloch Gösterimi (Bloch representation for one qubit mixed states)

Bölüm00 notlarında, tek bir kübitlik saf durumların Bloch küresi üzerindeki vektörler cinsinden ne kadar temiz bir geometrik gösterimi olduğunu görmüştük. Özellikle, herhangi bir saf kuantum durumu şu Bloch vektörü ile tanımlanabiliyordu:
$$\vec{r} = (\cos\phi \sin\theta, \, \sin\phi \sin\theta, \, \cos\theta)$$

Peki aynı geometrik yapı karma durumlar (mixed states) için de geçerli midir? Bu sorunun cevabı evet! Somut olarak, herhangi bir tek-kübitlik $\rho$ yoğunluk matrisini Pauli matrisleri cinsinden her zaman şu formda yazabiliriz:

$$\rho = \frac{1}{2}\left(\mathbb{I} + v_x X + v_z Z + v_y Y\right) \quad \text{--- (1.30)}$$

Burada $X, Y, Z$ daha önce karşılaştığımız Pauli matrisleridir. Eğer $\rho$ bir saf (pure) durum ise, $\vec{v} = (v_x, v_y, v_z)$ vektörü tam olarak zaten bildiğimiz $\vec{r}$ Bloch vektörünün ta kendisidir! 

*   **Saf Durumlar İçin:** Vektörün uzunluğu tam olarak 1'dir: $\|\vec{v}\|_2^2 = v_x^2 + v_y^2 + v_z^2 = 1$. (Vektör kürenin tam **yüzeyindedir**).
*   **Karma Durumlar İçin:** Vektörün uzunluğu 1'den küçük olabilir: $\|\vec{v}\|_2^2 \leq 1$. (Vektör kürenin **içindedir**).

Dolayısıyla, elimizdeki $\vec{v}$ vektörünün uzunluğu ($\|\vec{v}\|_2 \leq 1$), bize o matrisin geçerli bir tek-kübit kuantum durumu olup olmadığını anında söyler!


$\mathcal{P} = \{\mathbb{I}, X, Z, Y\}$ matris kümesinin, $2 \times 2$ boyutundaki tüm yoğunluk matrislerinin uzayı için bir **taban (basis)** oluşturduğunu fark etmek önemlidir. Bu matrislerin tamamının **Hilbert-Schmidt iç çarpımı** ($\langle A, B \rangle = \text{tr}(A^\dagger B)$) altında birbirine dik (orthogonal) olduğunu görebiliriz:

$$\text{tr}(X^\dagger Y) = \text{tr}(X^\dagger Z) = \text{tr}(X^\dagger \mathbb{I}) = 0 \quad \text{--- (1.31)}$$

Aynı diklik şartı, kümedeki diğer tüm farklı matris çiftleri için de tamamen geçerlidir.


> 🟦 **Alıştırma 1.2.4 ** Yukarıdaki (1.31) numaralı ortogonallik şartını ve Hadamard taban elemanlarının matris formlarını kullanarak şu eşitlikleri ispatlayınız:
> 
> $$|+\rangle\langle+| = \frac{1}{2}(\mathbb{I} + X) \quad \text{--- (1.32)}$$
> 
> $$|-\rangle\langle-| = \frac{1}{2}(\mathbb{I} - X) \quad \text{--- (1.33)}$$

**Çözüm:**
Ders notlarımızın önceki kısımlarında $|+\rangle\langle+|$ ve $|-\rangle\langle-|$ dış çarpım matrislerini şu şekilde bulmuştuk:
$$|+\rangle\langle+| = \frac{1}{2}\begin{pmatrix} 1 & 1 \\ 1 & 1 \end{pmatrix} \quad \text{ve} \quad |-\rangle\langle-| = \frac{1}{2}\begin{pmatrix} 1 & -1 \\ -1 & 1 \end{pmatrix}$$

Şimdi eşitliğin sağ tarafındaki Pauli matrisi açılımlarını yapalım ($\mathbb{I} = \begin{pmatrix} 1 & 0 \\ 0 & 1 \end{pmatrix}$ ve $X = \begin{pmatrix} 0 & 1 \\ 1 & 0 \end{pmatrix}$):
$$\frac{1}{2}(\mathbb{I} + X) = \frac{1}{2}\left[ \begin{pmatrix} 1 & 0 \\ 0 & 1 \end{pmatrix} + \begin{pmatrix} 0 & 1 \\ 1 & 0 \end{pmatrix} \right] = \frac{1}{2}\begin{pmatrix} 1 & 1 \\ 1 & 1 \end{pmatrix} = |+\rangle\langle+|$$

$$\frac{1}{2}(\mathbb{I} - X) = \frac{1}{2}\left[ \begin{pmatrix} 1 & 0 \\ 0 & 1 \end{pmatrix} - \begin{pmatrix} 0 & 1 \\ 1 & 0 \end{pmatrix} \right] = \frac{1}{2}\begin{pmatrix} 1 & -1 \\ -1 & 1 \end{pmatrix} = |-\rangle\langle-|$$

Böylece aradaki ilişki matris bazında da kusursuzca doğrulanmış olur.


> 🟦 **Alıştırma 1.2.2** $\mathcal{P}$ kümesindeki tüm farklı $M, N \in \mathcal{P}$ ($M \neq N$) matrislerinin birbiriyle anti-komüte ettiği, yani **$\{M,N\} = MN + NM = 0$** olduğu bilgisini kullanınız. Bu bilgiden yararlanarak, $M \neq N \in \mathcal{P}$ olduğu her durumda **$\text{tr}(MN) = 0$** olduğunu gösteriniz.


In [6]:
import numpy as np

# Matris tanımlamaları
I = np.eye(2)
X = np.array([[0, 1], [1, 0]])

# Alıştırma 1.2.4'ün sağ taraf hesaplamaları
plus_pure_teorik = 0.5 * (I + X)
minus_pure_teorik = 0.5 * (I - X)

print("1/2 * (I + X) Sonucu:\n", plus_pure_teorik)

# Hilbert-Schmidt iç çarpımı kontrolü: tr(X^\dagger * Y) = 0 mı?
Y = np.array([[0, -1j], [1j, 0]])
X_dagger = X.conj().T

ic_carpim = np.trace(np.dot(X_dagger, Y))
print(f"\nX ve Y matrislerinin iç çarpımı (iz değeri): {ic_carpim} (0 mı?: {np.isclose(ic_carpim, 0)})")


1/2 * (I + X) Sonucu:
 [[0.5 0.5]
 [0.5 0.5]]

X ve Y matrislerinin iç çarpımı (iz değeri): 0j (0 mı?: True)
